corehq/apps/es/case_search.py

```diff
@@ -0,0 +1,86 @@
+"""
+CaseSearchES
+------
+
+.. code-block:: python
+
+from corehq.apps.es import case_search as case_search_es
+
+    q = (case_search_es.CaseSearchES()
+         .domain('testproject')
+"""
+from . import filters, queries
+
+from corehq.apps.es.cases import CaseES
+from corehq.pillows.const import CASE_SEARCH_ALIAS
+
+
+PATH = "case_properties"
+
+
+class CaseSearchES(CaseES):
+    index = CASE_SEARCH_ALIAS
+
+    @property
+    def builtin_filters(self):
+        return [case_property_filter] + super(CaseSearchES, self).builtin_filters
+
+    @property
+    def _case_property_queries(self):
+        """
+        Returns all current case_property queries
+        """
+        try:
+            return self.es_query['query']['filtered']['query']['bool']['must']
+        except (KeyError, TypeError):
+            return []
+
+    def case_property_query(self, key, value, clause=queries.MUST, fuzzy=False):
+        """
+        Search for a case property.
+        Usage: (CaseSearchES()
+                .domain('swashbucklers')
+                .case_property_query("name", "rebdeard", "must", fuzzy=True)
+                .case_property_query("age", "15", "must")
+                .case_property_query("has_parrot", "yes", "should")
+                .case_property_query("is_pirate", "yes", "must_not"))
+
+        Can be chained with regular filters . Running a set_query after this will destroy it.
+        Clauses can be any of SHOULD, MUST, or MUST_NOT
+        """
+        fuzziness = "AUTO" if fuzzy else "0"
+        # Filter by case_properties.key and do a text search in case_properties.value
+        new_query = queries.nested(
+            PATH,
+            queries.filtered(
+                queries.match(value, "{}.value".format(PATH), fuzziness),
+                filters.term("{}.key".format(PATH), key)
+            )
+        )
+        return self._add_query(new_query, clause)
```

Comment:
This looks sufficiently complex that it might be nice to add a test to make sure it outputs the JSON you expect it to output.